### DISCUSSION TITLE
Extending XAML

### DISCUSSION THREAD

**Discussion by doublecount** [AvaloniaUI/Avalonia#12952] [Ideas]
What about extending XAML?
XAML is great, but sometimes I wish it would improve in some areas, evolve.

Here some ideas:

eg contructors: <TextBox(ctor1: xy, ctor2: xy)/>
eg generics: <NumInput+decimal/>
eg datatype: <Panel ¦ DataType/>
eg naming controls: <StackPanel - LeftPane/>

combined:
<StackPanel-LeftPane¦DataTypePerson] 
  <TextBox(bindTo:LastName)/>
  <TextBox(bindTo:FirstName)/>
  <NumInput+decimal(bindTo:Salary)/>
  <NumInput+int(bindTo:Age)/>
/>

Cheers!

**Comment by maxkatz6** [comment]
Please check this issue for some ideas https://github.com/AvaloniaUI/Avalonia/issues/2502
In general, there are no plans right now, but in general, we think Blazor syntax might work nicely as an alternative.

> eg contructors:

https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/xarguments-directive

> eg generics:

https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/generics

> eg datatype:

x:DataType ?

**Reply by doublecount** [reply]
the idea is to write XAML more readable
yes they work, but the x:arguments, x:datatype, x:TypeArguments are just 'ugly'

**Reply by cesarchefinho** [reply]
see levitalli compiled bindings

https://github.com/levitali/CompiledBindings

and

https://github.com/levitali/CompiledBindings/issues/28

**Reply by doublecount** [reply]
The levitalli markup extensions are nice! I have similar ones :)

In many cases, I do the bindingin code-behind though, with a single property and code behind that does everything.
Same thing, but hidden from XAML. 
The idea behind having a ctor for controls in XAML is to set the behavior at initialization (instead later on with properties). Problem with properties: some things only work in order, or half the code is to check for everyting in place all the time. 
Everyone probably has come accross one property needs to be set before another (there is such in Avalonia too).

Currently I mimic the ctors with having derived classes. But it would be so much nicer to create a XAML control with generic and ctors instead. Maybe sometime in future.
